src/tests/test_serialization.cpp

```diff
@@ -506,9 +506,7 @@ BOOST_AUTO_TEST_CASE( utils_unicode_test )
 	std::u32string water_u4;
 	water_u4.push_back(0x6C34);
 	std::string water_u8 = unicode_cast<std::string>(water_u4);
-	std::u16string water_u16 = unicode_cast<std::u16string>(water_u4);
 
-	BOOST_CHECK_EQUAL(water_u4[0], static_cast<char32_t>(water_u16[0]));
```

Comment:
Does Boost.UnitTest offer a way to customize how a type is streamed without defining an `operator<<`? I know Catch has a way to do this… if my guess is right and the only issue here is that the result of the test cannot be output to the stream due to non-existent operators, then I think the proper solution is to implement an alternate means of streaming it.

Reply:
They were added in 2014 in 1f4fe17576366c8371769c689d1eea62184cf657. In 2018 I changed the custom utf8 types to standard C++11 types (dfce371287c336e03ac85733420b22bbf283ad6b). It seems less important if it's using std::u32string et all to check that the implementation works. According to [P1423R3](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/p1423r3.html#option7), the behavior pre-C++20 when writing wide characters to stream was to output "the code units of the literals," which changed in C++20 due to whatever the hell they were doing with `char8_t`.

Reply:
The important thing is that Boost will no longer allow those checks as-is since it requires them to be writable to ostream. In C++20 the `ostream::operator<<` overloads for wide characters were explicitly deleted, so though it works with C++17, if we wanted to keep the tests we'd need to do something different. Which doesn't seem worth it.

Reply:
I found `BOOST_TEST_DONT_PRINT_LOG_VALUE` in [the documentation](https://www.boost.org/doc/libs/1_67_0/libs/test/doc/html/boost_test/utf_reference/testout_reference/test_output_macro_disable_type.html). It's not ideal, but maybe using that here would allow the deleted lines to compile on C++20?

Even better, it [looks like](https://www.boost.org/doc/libs/1_67_0/libs/test/doc/html/boost_test/test_output/test_tools_support_for_logging/testing_tool_output_disable.html) you can define `std::ostream& boost_test_print_type(std::ostream& ostr, ArgumentType const& right);` and it will take priority over `operator<<`. You could define it to cast the string to UTF-8 and output it, or you could define it to output it as a series of hex codes, or whatever you think works.

The problem with this change is that it's deleting a test, and as far as I can tell, the test itself is perfectly valid on C++20 – the issue is that the code to output whether the test passes or fails won't compile. We probably need the test (though I don't know for certain, but I presume it was added for a very good reason), so we should be fixing the output code instead of just deleting it.

Reply:
> Which doesn't seem worth it.

Maybe @AI0867 would have a better idea of just how much it's worth it, as the person who submitted the original commit. (No idea if he's around these days.)

Reply:
It seems reasonable to me to leave the check in, _but_ there is the fact that we're using _standard_ Unicode string types now, so I could maybe see an argument that it's unnecessary?

Reply:
If we want to do more work on this, I'll handle it separately, but for now it looks like removal is acceptable.